### PR TITLE
[BugFix] Different column types in different chunks

### LINKED ISSUE #22342: Broker load failure `type:LOAD_RUN_FAIL; msg:mismatched row count: 512 vs 4096`
### Steps to reproduce the behavior (Required)
1. Create a table in PRIMARY KEY with datetime key
```
CREATE TABLE `tbl_simple_pk` ( `key1` datetime not NULL COMMENT "", `key2` int NULL COMMENT "" ) ENGINE = OLAP PRIMARY KEY(`key1`) COMMENT "OLAP" DISTRIBUTED BY HASH(`key1`) BUCKETS 1 PROPERTIES ( "replication_num" = "3");
```
3. Load file via broker load
```
LOAD LABEL label10 ( DATA INFILE("hdfs://127.0.0.1:9000/4096_null/a.csv") INTO TABLE tbl_simple_pk COLUMNS TERMINATED BY "," FORMAT AS "csv" (key1,key2) SET (key1=key1,key2=key2)) WITH BROKER 'sr_broker' ;

```
The file includes records whose field is null, but the corresponding table columns is not null.

### Expected behavior (Required)
`type:ETL_QUALITY_UNSATISFIED; msg:quality not good enough to cancel` is returned. You can check the error detail in error URL `Error: NULL value in non-nullable column 'key1'. Row: [NULL, 3, 0]`.

```
         JobId: 12020
         Label: label10
         State: CANCELLED
      Progress: ETL:N/A; LOAD:N/A
          Type: BROKER
      Priority: NORMAL
       EtlInfo: unselected.rows=0; dpp.abnorm.ALL=1; dpp.norm.ALL=4097
      TaskInfo: resource:N/A; timeout(s):14400; max_filter_ratio:0.0
      ErrorMsg: type:ETL_QUALITY_UNSATISFIED; msg:quality not good enough to cancel
    CreateTime: 2023-04-24 10:48:42
  EtlStartTime: 2023-04-24 10:48:43
 EtlFinishTime: 2023-04-24 10:48:43
 LoadStartTime: 2023-04-24 10:48:43
LoadFinishTime: 2023-04-24 10:48:43
           URL: http://127.0.0.1:8064/api/_load_error_log?file=error_log_b2f2dacd78884185_a41f521e8b6e4b0a
    JobDetails: {"Unfinished backends":{"b2f2dacd-7888-4185-a41f-521e8b6e4b08":[]},"ScannedRows":4098,"TaskNumber":1,"All backends":{"b2f2dacd-7888-4185-a41f-521e8b6e4b08":[10002]},"FileNumber":1,"FileSize":16394}
```
### Real behavior (Required)
`type:LOAD_RUN_FAIL; msg:mismatched row count: 512 vs 4096` is retuned. And the error URL is `NULL`.
```
         JobId: 12043
         Label: label10
         State: CANCELLED
      Progress: ETL:N/A; LOAD:N/A
          Type: BROKER
      Priority: NORMAL
       EtlInfo: NULL
      TaskInfo: resource:N/A; timeout(s):14400; max_filter_ratio:0.0
      ErrorMsg: type:LOAD_RUN_FAIL; msg:mismatched row count: 512 vs 4096
    CreateTime: 2023-04-24 10:57:14
  EtlStartTime: 2023-04-24 10:57:18
 EtlFinishTime: 2023-04-24 10:57:18
 LoadStartTime: 2023-04-24 10:57:18
LoadFinishTime: 2023-04-24 10:57:18
           URL: NULL
    JobDetails: {"Unfinished backends":{"a1550bfc-e398-4d24-be77-7a27b2720947":[]},"ScannedRows":0,"TaskNumber":1,"All backends":{"a1550bfc-e398-4d24-be77-7a27b2720947":[10002]},"FileNumber":1,"FileSize":188491}
```
### StarRocks version (Required)
 2.4.4

### PR DESCRIPTION
**Pull request by rickif**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/22342

## Problem Summary(Required) ：

The column type like datetime is cast from varchar in the file scanner.
cast_from_string_to_datetime_fn would build a column as nullable when there're nulls in the column, while it would build a column as non-nullable when there's no null in the column. So, the column types of different chunks may be different.
The exchange node gets the chunk schema according to the first chunk it receives and then parses the chunk according to the schema. In this way, the exchange node cannot handle the variable column type.
This PR corrects the columns returned by the cast_from_string_to_datetime_fn according to the actual schema.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/vectorized/file_scanner.cpp [hunk 1]
```diff
@@ -0,0 +1,317 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "exec/vectorized/file_scanner.h"
```
Comment (trueeyu) [inline be/src/exec/vectorized/file_scanner.cpp:3]: Why add the file?

Comment (rickif) [reply be/src/exec/vectorized/file_scanner.cpp]: Already removed.

**Review by trueeyu — DISMISSED** [review]

File: be/src/exec/file_scanner.cpp [hunk 2]
```diff
@@ -185,11 +185,10 @@ StatusOr<ChunkPtr> FileScanner::materialize(const starrocks::ChunkPtr& src, star
 
         col = ColumnHelper::unfold_const_column(slot->type(), cast->num_rows(), col);
 
-        // The column builder in ctx->evaluate may build column as non-nullable.
+        // The column builder in ctx->evaluate may build column as non-nullable/nullable,
+        // which is differenet from expected schema.
```
Comment (chaoyli) [inline be/src/exec/file_scanner.cpp:189]: spelling error, different

Comment (rickif) [reply be/src/exec/file_scanner.cpp]: Already corrected.

**Comment by rickif** [comment]
run all

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by rickif** [comment]
@Mergifyio rebase

**Review by trueeyu — DISMISSED** [review]

File: be/src/exec/file_scanner.cpp [hunk 3]
```diff
@@ -212,26 +204,56 @@ StatusOr<ChunkPtr> FileScanner::materialize(const starrocks::ChunkPtr& src, star
                     filter[i] = 0;
                     _error_counter++;
 
+                    std::stringstream error_msg;
```
Comment (meegoo) [inline be/src/exec/file_scanner.cpp:207]: why move it out of if statement? this causes unnecessary overhead.

**Review by meegoo — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.4

**Pull request merged**